keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,955 @@
+   - https://github.com/kubernetes/kubernetes/issues/119645
+   - https://github.com/kubernetes/kubernetes/issues/125579
+   - https://github.com/kubernetes/kubernetes/issues/127793
+
+
+Hence, it is necessary to handle capacity updates gracefully across the cluster, rather than resetting the cluster components to achieve the same outcome.
+
+Also, given that the capability to live-resize a node exists in the Linux and Windows kernels, enabling the kubelet to be aware of the underlying changes in the node's compute capacity will mitigate any actions that are required to be made
+by the Kubernetes administrator.
+
+Node resource hot plugging proves advantageous in scenarios such as:
+- Efficiently managing resource demands with a limited number of nodes by increasing the capacity of existing nodes instead of provisioning new ones, which brings less overhead on the control-plane.
+- The procedure of establishing new nodes is considerably more time-intensive than expanding the capabilities of current nodes.
+- Improved inter-pod network latencies as the inter-node traffic can be reduced if more pods can be hosted on a single node.
+- Mitigate a few of the existing limitations/issues that are associated with a node/kubelet restart.
+
+Implementing this KEP will empower nodes to recognize and adapt to changes in their compute configurations and allow facilitating the efficient and effective deployment of pod workloads to nodes capable of meeting the required compute demands.
+
+### Goals
+
+* Achieve seamless node capacity expansion through resource hotplug.
+* Enable the re-initialization of resource managers (CPU manager, memory manager) and kube runtime manager without reset to accommodate alterations in the node's resource allocation.
+* Recalculating and updating swap memory limit for existing pods.
+
+### Non-Goals
+
+* Dynamically adjust system reserved and kube reserved values.
+* Hot unplug of node resources.
+* Update the autoscaler to utilize resource hotplug.
+* Re-balance workloads across the nodes.
+* Update runtime/NRI plugins with host resource changes.
+
+## Proposal
+
+This KEP strives to enable node resource hot plugging by making the kubelet to watch and retrieve machine resource information from cAdvisor's cache as and when it changes, cAdvisor's cache is already updated periodically.
+The kubelet will fetch this information, subsequently entrusting the node status updater to disseminate these updates at the node level across the cluster.
+Moreover, this KEP aims to refine the initialization and reinitialization processes of resource managers, including the memory manager and CPU manager, to ensure their adaptability to changes in node configurations.
+With this proposal it's also necessary to recalculate and update OOMScoreAdj and swap limit for the pods that had been existing before resize. But this carries a small overhead due to recalculation of swap and OOMScoreAdj.
```

Comment:
is this still accurate?

Reply:
OOMScoreAdj will not be recalculated. Right @Karthik-K-N?

Reply:
Yeah, OOMScoreAdj is not recalculated, I just updated it. Thank you.